src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,142 @@
+import * as classNames from 'classnames';
+import * as PropTypes from 'prop-types';
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+
+import Button, { ButtonProps } from '../Button/Button';
+import * as Rivet from '../Rivet';
+import ModalEvent from './ModalEvent';
+
+
+export interface ModalProps {
+    /**
+     * Determines whether the modal is shown or not
+     */
+    isOpen?: boolean;
+    /**
+     * Optional function to call to raise when the alert is dismissed. If undefined, the modal will act as a dialog (i.e., no close button, 
+     * will not close on outside click or escape key)
+     */
+    onDismiss?: Rivet.Action;
+    /**
+     * The content of the modal's header
+     */
+    title: React.ReactNode | string;
+    /**
+     * Button or buttons which will appear in the footer of the modal
+     */
+    controls: React.ReactElement<ButtonProps> | Array<React.ReactElement<ButtonProps>>;
```

Comment:
I personally don't like passing arrays of React elements as properties.  Have you considered using components to denote what goes in the header/body/footer like other libraries use?

Reply:
On top of this it feels odd to be that we allow passing arbitrary controls then render and additional one if an `onDismiss` handler is provided.

Reply:
I'm not a fan of passing in arrays of elements either, but this was what was started before I began. I can change it.

As far as the "close" button, it is not in the control section at the bottom, it is in the upper right corner (confusing as its at the bottom of the component definition). Having it depend on a function being passed in is similar to what react-boostrap does, except they have an additional property (`closeButton`) and describe their `onHide` as "A callback fired when the header closeButton or non-static backdrop is clicked. Required if either are specified." Essentially, the current implementation says if there is a function to handle dismissing, you get a close button for free, as opposed to react-bootstrap, that has an additional prop that you can toggle *only if you pass in the function*, so both render the close button only if the close handler function is provided.

Reply:
Looking at separating out `ModalHeader` etc., we are currently setting an `aria-labelledby` in the parent that matches the `id` of the header. Would this best practice be delegated to the user or how would we enforce it?

Reply:
Yeah, I did the original control button array implementation before I left mostly as a way to quickly reuse an existing component and enforce the type, but I have no qualms about nested components for the controls. Also, what Andy said about the close button: it's confusing, but that's the way the Rivet documentation renders dismissible modals (as well as how react-bootstrap does it, as Andy pointed out).

Reply:
I'm thinking we could have the header baked into the Modal (to resolve the issue I mentioned above) but then use `ModalBody` and `ModalFooter` children to more closely match `react-bootstrap`. Thoughts?

Reply:
Also, I haven't pushed it yet, but I am moving the "close" button into the `header` tag to make it more apparent  where it will actually render